kafka-rest/src/main/java/io/confluent/kafkarest/Errors.java

```diff
@@ -63,6 +63,14 @@ public static RestException leaderNotAvailableException() {
     return new RestNotFoundException(LEADER_NOT_AVAILABLE_MESSAGE, LEADER_NOT_AVAILABLE_ERROR_CODE);
   }
 
+  public static final String INVALID_PAYLOAD_MESSAGE = "Payload error. ";
+  public static final int INVALID_PAYLOAD_ERROR_CODE = 40405;
+
+  public static RestConstraintViolationException invalidPayloadException(String cause) {
```

Comment:
I updated the error code to 404, but all 404 error finally throws `RestException`, I'm not sure we should throw a RestConstraintViolationException or RestException. @ehumber @dimitarndimitrov any suggestions for this?

Reply:
I don't think the exception type matters too much, as long as what the end user sees in the message text is useful to them.  Do they still see the "Request body is empty. Data is required." message even with the RestException?

Reply:
Yes, they will see the Request body is empty. Data is required. message

Reply:
If we change the error code to 404, it will cause `ProduceActionTest.produceNoSchemaRegistryDefined` failed

```
14:07:29  [ERROR] Failures: 
14:07:29  [ERROR]   ProduceActionTest.produceNoSchemaRegistryDefined:171 
14:07:29    Expectation failure on verify:
14:07:29      ChunkedOutput.write(ErrorHolder{error=ErrorResponse{errorCode=404, message=Error: 40405 : Payload error. Schema Registry must be configured when using schemas.}}): expected: 1, actual: 0
14:07:29    Unexpected method calls:
14:07:29      ChunkedOutput.write(ErrorHolder{error=ErrorResponse{errorCode=422, message=Error: 40405 : Payload error. Schema Registry must be configured when using schemas.}})
14:07:29  [INFO] 
14:07:29  [ERROR] Tests run: 521, Failures: 1, Errors: 0, Skipped: 5
```
RestConstraintViolationException can only return 422 : https://github.com/confluentinc/rest-utils/blob/master/core/src/main/java/io/confluent/rest/exceptions/ConstraintViolationExceptionMapper.java#L31

But for 404, looks it's `Not found`, is it expected if no request body?

```
        PAYMENT_REQUIRED(402, "Payment Required"),
        FORBIDDEN(403, "Forbidden"),
        NOT_FOUND(404, "Not Found"),
        METHOD_NOT_ALLOWED(405, "Method Not Allowed"),
        NOT_ACCEPTABLE(406, "Not Acceptable"),
        PROXY_AUTHENTICATION_REQUIRED(407, "Proxy Authentication Required"),
        REQUEST_TIMEOUT(408, "Request Timeout"),
        CONFLICT(409, "Conflict"),
```

Reply:
So this test case has a -d payload passed in. That payload has a schema defined, but the proxy is configured to not point at a schema registry.

So it's not that there is no request body, it's that there is a request body that doesn't work for the particular configuration of the REST server.

404 not found doesn't sound quite right for that, I think 422 is better.

Can you make all the exception mappers line up to do that?  If not, shall we just get them to return the 422 like you coded this change to do to start with?

Reply:
Thanks @ehumber , I changed it to 422.

Currently I met jenkins error, I don't think they related to my change because I also saw them on my local with a clean repo. I will also take a look to see how to fix it.
```
21:57:08  [ERROR] Failures: 
21:57:08  [ERROR]   AvroProduceConsumeTest>SchemaProduceConsumeTest.produceThenConsume_returnsExactlyProduced:181->SchemaProduceConsumeTest.assertMapEquals:220 Expected and actual are not equal. Extra: [], Missing: [1, 2, 3], Different: []
21:57:08  [ERROR]   JsonSchemaProduceConsumeTest>SchemaProduceConsumeTest.produceThenConsume_returnsExactlyProduced:181->SchemaProduceConsumeTest.assertMapEquals:220 Expected and actual are not equal. Extra: [], Missing: [1, 2, 3], Different: []
21:57:08  [ERROR]   ProtobufProduceConsumeTest>SchemaProduceConsumeTest.produceThenConsume_returnsExactlyProduced:181->SchemaProduceConsumeTest.assertMapEquals:220 Expected and actual are not equal. Extra: [], Missing: [{"key":2}, {"key":3}, {"key":1}], Different: []
21:57:08  [INFO] 
21:57:08  [ERROR] Tests run: 180, Failures: 3, Errors: 0, Skipped: 1
```

Reply:
Yup, they are some of the flakey tests.  They should pass intermittently though, and on my laptop they pass very much more than they fail.  Is it a solid failure now?

Reply:
Yeah, solid failure now, even fail without any change.

Reply:
Did this get fixed?

Reply:
Yeah, this is fixed.

Reply:
I just run again on my local, let me run it on jenkins again.